go-controller/pkg/ovn/controller/admin_network_policy/admin_network_policy.go

```diff
@@ -32,6 +49,888 @@ func (c *Controller) processNextANPWorkItem(wg *sync.WaitGroup) bool {
+				libovsdbops.ObjectNameKey: currentANPState.name,
+			})
+		predicateFunc := func(as *nbdb.AddressSet) bool {
+			asPriority, _ := strconv.Atoi(as.ExternalIDs[libovsdbops.PriorityKey.String()])
+			return (as.ExternalIDs[libovsdbops.PolicyDirectionKey.String()] == ANPEgressPrefix &&
+				(len(desiredANPState.egressRules) == 0 || int32(asPriority) < desiredANPState.egressRules[len(desiredANPState.egressRules)-1].priority)) ||
+				(as.ExternalIDs[libovsdbops.PolicyDirectionKey.String()] == ANPIngressPrefix &&
+					(len(desiredANPState.ingressRules) == 0 || int32(asPriority) < desiredANPState.ingressRules[len(desiredANPState.ingressRules)-1].priority))
+		}
+		asPredicate = libovsdbops.GetPredicate[*nbdb.AddressSet](predicateIDs, predicateFunc)
+	}
+	if asPredicate != nil {
+		ops, err = libovsdbops.DeleteAddressSetsWithPredicateOps(c.nbClient, ops, asPredicate)
+		if err != nil {
+			return nil, fmt.Errorf("failed to create address-set destroy ops for ANP %s, err: %v", currentANPState.name, err)
+		}
+	}
+	// now CreateOrUpdate the address-sets; add the right IPs
+	for _, rule := range desiredANPState.ingressRules {
+		asIndex := GetANPPeerAddrSetDbIDs(desiredANPState.name, rule.gressPrefix, fmt.Sprintf("%d", rule.priority), c.controllerName)
+		addrSet, addrSetOps, err := c.addressSetFactory.NewAddressSetOps(asIndex, stringsToIPs(rule.podIPs.UnsortedList()))
+		if err != nil {
+			return nil, fmt.Errorf("failed to create address-sets for ANP %s's"+
+				" ingress rule %s/%s/%d: %v", desiredANPState.name, rule.name, rule.gressPrefix, rule.priority, err)
+		}
+		ops = append(ops, addrSetOps...)
+		rule.addrSet = addrSet
+	}
+	for _, rule := range desiredANPState.egressRules {
+		asIndex := GetANPPeerAddrSetDbIDs(desiredANPState.name, rule.gressPrefix, fmt.Sprintf("%d", rule.priority), c.controllerName)
+		addrSet, addrSetOps, err := c.addressSetFactory.NewAddressSetOps(asIndex, stringsToIPs(rule.podIPs.UnsortedList()))
+		if err != nil {
+			return nil, fmt.Errorf("failed to create address-sets for ANP %s's"+
+				" egress rule %s/%s/%d: %v", desiredANPState.name, rule.name, rule.gressPrefix, rule.priority, err)
+		}
+		ops = append(ops, addrSetOps...)
+		rule.addrSet = addrSet
+	}
+	return ops, nil
+}
+
+// constructOpsForPeerChanges takes the desired and current rules of the anp and returns the corresponding ops for updating NBDB objects
+func constructOpsForPeerChanges(desiredRules, currentRules []*gressRule, atLeastOneRuleUpdated *bool, anpName string) ([]ovsdb.Operation, error) {
+	var ops []ovsdb.Operation
+	for i := range desiredRules {
+		desiredRule := desiredRules[i]
+		currentRule := currentRules[i]
+		ipsToAdd := desiredRule.podIPs.Difference(currentRule.podIPs)
+		if len(ipsToAdd) > 0 {
+			klog.Infof("Adding podIPs %+v to address-set %s for ANP %s", ipsToAdd, desiredRule.addrSet.GetName(), anpName)
+			addrOps, err := desiredRule.addrSet.AddIPsReturnOps(stringsToIPs(ipsToAdd.UnsortedList()))
```

Comment:
From a reconciling perspective, I think that if we are aware of all the IPs the AS should have, its safer to just update the AS and set all the IPs, rather than mutating to add or remove. This comment applies to the PG ports as well.

Reply:
hmm I thought let's say if we had 500 IPs on the address-set, adding 2 IPs and removing 1IP is less heavier than sending an update op for setting 501IPs - but you raise a good point - which of these two types of ops is more bulky?

Reply:
Yeah, you are right, the op will be more bulky and we will pay a price when marshalling. The benefit is we would reconcile the data so kinda self healing capabilities if we are in a bad state previously.

Reply:
ack I acknowledge your point, since we have gone through the trouble of maintaining the caches for these things, might as well use it to their advantage i guess :)